### PR TITLE
Refactoring of azure obj store #3957 

### PR DESCRIPTION
**Pull request by wiardvanrij**
Signed-off-by: Wiard van Rij <wiard@outlook.com>

* [x] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

- Adds configuration options for the Azure Pipeline method
- Adds configuration options for the Azure Reader method
- Adds configuration options for the HTTP transport method
- Adds HTTP Transport override
- Adds MSI authentication (#3889)
- Made parsing of the config strict
- ~~Breaks configuration by moving `max_retries` to it's own `reader_config`~~
- Rewrote tests 

## Verification

**How you tested it?** 
No, other than the tests included

Especially the MSI authentication part is untested. 

**How do you know it works?**
This is tested via other users on https://github.com/thanos-io/thanos/issues/3957 : 

@wiardvanrij  ran your image all night locally with no failures.

_Originally posted by @airkewld in https://github.com/thanos-io/thanos/issues/3957#issuecomment-830087240_

---------
I hope this PR can get tested by users of Azure to confirm the working of the changes AND if this has positive impact for on, for example #3952 - relates to #3957

### CODE REVIEW

**Comment by wiardvanrij** [comment]
Need to check why the test fails on the cortex dep. :D

**Comment by wiardvanrij** [comment]
We use this part; https://github.com/cortexproject/cortex/blob/master/pkg/storage/bucket/azure/bucket_client.go#L10 and Cortex uses the Thanos part. 
I think for this change both Cortex and Thanos need to be updated but I'm not sure on how to proceed with this. 

@pracucci  could you weight in here if my statement is even correct? Thank you!

**Comment by bwplotka** [comment]
Cycling dependency problem, so:

* Fork Cortex, fix this on your fork, push remote branch to your fork
* Update go.mod to replace cortex dep with your fork URL and Commit SHA
* Let's review AND merge this PR
* Propose the change in the fork to Cortex 
* Remove replace.

File: pkg/objstore/azure/azure.go
Comment (bwplotka) [inline pkg/objstore/azure/azure.go:41]: Can we reuse this struct?

File: pkg/objstore/azure/helpers.go
Comment (bwplotka) [inline pkg/objstore/azure/helpers.go:119]: That sounds like something we can reuse across buckets.

**Review by bwplotka — COMMENTED** [review]
Some suggestions otherwise look good!

**Comment by phillebaba** [comment]
Is there anything I can do to help move this along? :) Or is this dependent on changes that have to be made in Cortex?

**Comment by wiardvanrij** [comment]
> Is there anything I can do to help move this along? :) Or is this dependent on changes that have to be made in Cortex?

It's currently not blocking but just needs some fixes in packages. Which eventually also requires some changes at the Cortex side. I did however not foresee the complexity at first :p 

But it's good to know that people need this, so I can make it a higher prio!

**Comment by airkewld** [comment]
Glad to see this moving along. My compact pod is dying multiple times a day at this point, hopefully this will resolve the issue for me and others.  
Thanks

File: pkg/objstore/azure/helpers.go
Comment (wiardvanrij) [reply pkg/objstore/azure/helpers.go:137]: ignore

**Comment by wiardvanrij** [comment]
@airkewld  & @phillebaba see https://github.com/thanos-io/thanos/issues/3957 - I've build an image. But feel free to build it yourself. Wouldn't mind more testers on this feature :)

**Comment by phillebaba** [comment]
I can probably test this as I am already running Thanos in AKS. Hopefully this will work with aad-pod-identity.

**Comment by wiardvanrij** [comment]
> I can probably test this as I am already running Thanos in AKS. Hopefully this will work with aad-pod-identity.

Don't use it on production yet tho.. :D Also can we move potential discussions to the issue? Makes it a bit easier. Thanks so much btw :)

**Comment by wiardvanrij** [comment]
> Cycling dependency problem, so:
> 
> * Fork Cortex, fix this on your fork, push remote branch to your fork
> * Update go.mod to replace cortex dep with your fork URL and Commit SHA
> * Let's review AND merge this PR
> * Propose the change in the fork to Cortex
> * Remove replace.

I changed the configuration in a way that it's not breaking anymore AND it has no dependency issue anymore. 
There is a technical trade-off here by implementing a configuration value that is somewhat obsolete. 

We now use the `max_retries` config value if no specific other 'retry related' configuration is set. This means there is 0 change for those we do not use the extra added configuration options. If one sets specific retry value configurations, the `max_retries` config value is ignored in favor of the more specific config values.  

So this resolved two things:
- No dependency issue anymore
- Backwards compatible

File: pkg/objstore/azure/helpers.go
Comment (wiardvanrij) [reply pkg/objstore/azure/helpers.go:119]: I would like to propose to not do this in this PR but create a new issue for this. As this would mean a small refactor in various parts in Thanos. This makes the scope of this PR too big IMO. WDYT?

File: pkg/objstore/azure/azure.go
Comment (wiardvanrij) [reply pkg/objstore/azure/azure.go:40]: Yes, we should. See my other comments about not doing it in this PR, but create a new issue/PR for this

**Comment by nicolastakashi** [comment]
Hi dudes!
any prevision to release it?

File: pkg/objstore/azure/azure.go
Comment (yashrsharma44) [inline pkg/objstore/azure/azure.go:150]: unless I have misunderstood, do you mean 
```suggestion
	// This part backwards compatibility but also ease of configuration
```
😅

**Review by yashrsharma44 — COMMENTED** [review]
Small question about the comment 😛

File: pkg/objstore/azure/azure.go
Comment (bwplotka) [reply pkg/objstore/azure/azure.go:40]: We could, but why? (:

**Review by bwplotka — DISMISSED** [review]
LGTM, 💪🏽

File: pkg/objstore/azure/azure.go
Comment (wiardvanrij) [reply pkg/objstore/azure/azure.go:150]: Yes, thanks, I will make this proper English :D

**Comment by wiardvanrij** [comment]
I will update this PR as soon as this is merged: https://github.com/thanos-io/thanos/pull/4104#issuecomment-831352483
Then I can use that generic http method 💯

**Comment by phillebaba** [comment]
@wiardvanrij had some time now to check this PR. The container is created without an issue but I am seeing authorization issues when uploading blocks from the receiver. I am using aad-pod-identity which adds additional complexities, but unless someone else has verified that this feature works I think you should hold off on merging.

**Comment by wiardvanrij** [comment]
> @wiardvanrij had some time now to try this PR. The container is created without an issue but I am seeing authorization issues when uploading blocks from the receiver. I am using aad-pod-identity which adds additional complexities, but unless someone else has verified that this feature works I think you should hold off on merging.

Thanks! Could you perhaps provide more information in https://github.com/thanos-io/thanos/issues/3957 ?

File: docs/storage.md
Comment (phillebaba) [inline docs/storage.md:355]:
```suggestion
If `msi_resource` is used, authentication is done via ServicePrincipalToken. The value for Azure should be `https://<storage-account-name>.blob.core.windows.net`.
```

File: pkg/objstore/azure/azure_test.go
Comment (phillebaba) [inline pkg/objstore/azure/azure_test.go:119]:
```suggestion
msi_resource: "https://example.blob.core.windows.net"
```

File: pkg/objstore/azure/azure_test.go
Comment (phillebaba) [inline pkg/objstore/azure/azure_test.go:133]:
```suggestion
msi_resource: "https://example.blob.core.windows.net"
```

**Review by phillebaba — COMMENTED** [review]
Just small suggestions to fix the msi resource value.

File: pkg/objstore/azure/azure.go
Comment (bwplotka) [reply pkg/objstore/azure/azure.go:40]: Cool let's add in https://github.com/thanos-io/thanos/issues/4104 or so

File: pkg/objstore/azure/helpers.go
Comment (bwplotka) [inline pkg/objstore/azure/helpers.go:49]:
```suggestion
		// Get a new token.
```

File: pkg/objstore/azure/helpers.go
Comment (bwplotka) [inline pkg/objstore/azure/helpers.go:40]:
```suggestion
```

File: pkg/objstore/azure/helpers.go
Comment (bwplotka) [inline pkg/objstore/azure/helpers.go:38]:
```suggestion
```

File: pkg/objstore/azure/helpers.go
Comment (bwplotka) [reply pkg/objstore/azure/helpers.go:40]: no need for those new lines

File: pkg/objstore/azure/helpers.go
Comment (bwplotka) [inline pkg/objstore/azure/helpers.go:57]:
```suggestion
```

File: pkg/objstore/azure/helpers.go
Comment (bwplotka) [inline pkg/objstore/azure/helpers.go:65]:
```suggestion
```

File: pkg/objstore/azure/helpers.go
Comment (bwplotka) [inline pkg/objstore/azure/helpers.go:58]: No need for this else we return above anyway. See https://thanos.io/tip/contributing/coding-style-guide.md/#control-structure-prefer-early-returns-and-avoid-else

**Review by bwplotka — APPROVED** [review]
LGTM! There are minor style nits, otherwise I would be happy to merge this.

Great work and sorry for major lag on review!

File: pkg/objstore/azure/helpers.go
Comment (wiardvanrij) [reply pkg/objstore/azure/helpers.go:56]: Removed the `else` so this newline is now legit.

File: go.mod
Comment (wiardvanrij) [reply go.mod:8]: https://github.com/Azure/go-autorest/pull/640

**Review by KingTropea999 — APPROVED** [review]

File: go.mod
Comment (phillebaba) [reply go.mod:8]: @wiardvanrij v0.5.8 should now be out with an updated /x/crypto dependency.

File: docs/storage.md
Comment (tomasfreund) [inline docs/storage.md:346]: Could [tls_config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tls_config) be added instead of just `insecure_skip_verify`?
I know blob storage itself does not support it but we'd like to use a proxy (by changing the `endpoint` in the configuration) with mTLS for blob storage to avoid having to provide the storage account key to every cluster we monitor.
I also think it would be better to make the config as similar to Prometheus as possible for consistency and code resuse (if possible)

Comment (wiardvanrij) [reply docs/storage.md:346]: yes and no. This is in line with all other storage options in Thanos: https://thanos.io/tip/thanos/storage.md/
There is some work being done to make these configs more logical, but for now I try to keep it aligned with how everything in Thanos works (:

File: go.mod
Comment (wiardvanrij) [reply go.mod:8]: I will look at this monday, as I'm on vacation. Feel free to make a PR against my branch tho :)

File: docs/storage.md
Comment (tomasfreund) [reply docs/storage.md:346]: thanks for the answer, I guess the solution to this problem is to implement a custom proxy and run it as a sidecar, using https with a dummy certificate and `insecure_skip_verify`

Comment (wiardvanrij) [reply docs/storage.md:346]: > thanks for the answer, I guess the solution to this problem is to implement a custom proxy and run it as a sidecar, using https with a dummy certificate and `insecure_skip_verify`

Yea I understand the case, it's just that I don't feel making one specific change here. However, it would make sense to implement a generic tls_config for non-component related configs, but for 'object storage connections'. As this would/should be implemented not only for Azure Blob, but also for all the other options Thanos supports. 
Perhaps you could create a new issue for this?

**Review by bwplotka — APPROVED** [review]
Amazing! Let's get this in and get some nice, sweet feedback.

Nice work @wiardvanrij , Mr Golang dev! (:

**Comment by amrmahdi** [comment]
@wiardvanrij Thanks for adding these changes! 
Couple of questions:
1. The `msi_resource` why is it needed? Isn't it just a concat of storage account name and the endpoint?
2. I'm assuming this does not support user assigned managed identity, correct?

**Comment by amrmahdi** [comment]
> @wiardvanrij Thanks for adding these changes!
> Couple of questions:
> 
> 1. The `msi_resource` why is it needed? Isn't it just a concat of storage account name and the endpoint?
> 2. I'm assuming this does not support user assigned managed identity, correct?

I think got my questions answered. I have a follow up PR to support user-assigned managed identity. https://github.com/thanos-io/thanos/pull/4636

**Comment by wiardvanrij** [comment]
> > @wiardvanrij Thanks for adding these changes!
> > Couple of questions:
> > 
> > 1. The `msi_resource` why is it needed? Isn't it just a concat of storage account name and the endpoint?
> > 2. I'm assuming this does not support user assigned managed identity, correct?
> 
> I think got my questions answered. I have a follow up PR to support user-assigned managed identity. #4636

Wow thanks! Really good and thanks for helping out!

**Pull request merged**
